### ISSUE TITLE
Include Deactivate Listing in Job Actions

### ISSUE THREAD

**Issue by workpei25** [Automattic/WP-Job-Manager#2860]
### Is your feature request related to a problem? Please describe
Employers often need to remove job listings for various reasons, but the only option they have is to Delete a job from their jobs dashboard.

### Describe the solution you'd like
Include an action in the jobs dashboard menu that allows employers to deactivate a job listing (sending the listing to Draft status).

**Comment by Braehler** [comment]
@Scotm19 
this is a long awaited function but still not included. I recently got this to work. Just put this into your functions.php 

`<?php
/**
 * Function to add deactivate Button toJob Dashboard
 * Works with WP Job Manager 2.4.0 and WC Paid Listings 3.0.3
 */

// add Button 
add_action('job_manager_my_job_actions', 'custom_add_deactivate_button', 10, 2);
function custom_add_deactivate_button($actions, $job) {
    if ($job->post_status === 'publish') {
        // Basie-URL Dashboards
        $dashboard_url = get_permalink(get_option('job_manager_job_dashboard_page_id'));
        
        // URLwith correct parameters
        $actions['deactivate'] = array(
            'label' => __('Deactivate', 'wp-job-manager'),
            'url'   => add_query_arg(array(
                'action' => 'deactivate',
                'job_id' => $job->ID
            ), $dashboard_url)
        );
    }
    return $actions;
}

// prepare deactivate
add_action('wp', 'custom_process_job_deactivation');
function custom_process_job_deactivation() {
    if (!isset($_GET['action']) || $_GET['action'] !== 'deactivate') {
        return;
    }

    if (empty($_GET['job_id'])) {
        return;
    }

    $job_id = absint($_GET['job_id']);

    // check permission
    $job = get_post($job_id);
    if (!$job || $job->post_author != get_current_user_id()) {
        wp_die(__('You do not have the permission to deactivate', 'wp-job-manager'));
    }

    // update Job Status
    wp_update_post(array(
        'ID' => $job_id,
        'post_status' => 'expired'
    ));

    // WC Paid Listings 
    if (class_exists('WC_Paid_Listings')) {
        delete_post_meta($job_id, '_package_id');
        delete_post_meta($job_id, '_package_subscription_id');
        delete_post_meta($job_id, '_featured');
    }

    // clear Cache
    clean_post_cache($job_id);

    // redirect to job dashboard
    wp_redirect(add_query_arg('msg', 'deactivated', get_permalink(get_option('job_manager_job_dashboard_page_id'))));
    exit;
}

// success message
add_action('job_manager_job_dashboard_before', 'custom_show_deactivation_notice');
function custom_show_deactivation_notice() {
    if (isset($_GET['msg']) && $_GET['msg'] === 'deactivated') {
        echo '<div class="job-manager-message">' . 
             esc_html__('Success.', 'wp-job-manager') . 
             '</div>';
    }
}`

**Comment by Ericmuc** [comment]
hey, great, thanks, I have adopted it and it works just fine, it should find its way into the core, very helpful.

Here the WPML adapted version, remains in the current language:

/**
 * Function to add deactivate Button to Job Dashboard
 * Works with WP Job Manager 2.4.0 and WC Paid Listings 3.0.3
 */
// Add Deactivate Button
add_action('job_manager_my_job_actions', 'custom_add_deactivate_button', 10, 2);
function custom_add_deactivate_button($actions, $job) {
    if ($job->post_status === 'publish') {
        // Base URL Dashboard
        $dashboard_url = get_permalink(get_option('job_manager_job_dashboard_page_id'));

        // Add Button with correct parameters
        $actions['deactivate'] = array(
            'label' => __('Deactivate', 'wp-job-manager'),
            'url'   => add_query_arg(array(
                'action' => 'deactivate',
                'job_id' => $job->ID
            ), $dashboard_url)
        );
    }
    return $actions;
}

// Prepare Deactivate
add_action('wp', 'custom_process_job_deactivation');
function custom_process_job_deactivation() {
    if (!isset($_GET['action']) || $_GET['action'] !== 'deactivate') {
        return;
    }

    if (empty($_GET['job_id'])) {
        return;
    }

    $job_id = absint($_GET['job_id']);

    // Check Permission
    $job = get_post($job_id);
    if (!$job || $job->post_author != get_current_user_id()) {
        wp_die(__('You do not have the permission to deactivate', 'wp-job-manager'));
    }

    // Update Job Status
    wp_update_post(array(
        'ID' => $job_id,
        'post_status' => 'expired'
    ));

    // WC Paid Listings cleanup
    if (class_exists('WC_Paid_Listings')) {
        delete_post_meta($job_id, '_package_id');
        delete_post_meta($job_id, '_package_subscription_id');
        delete_post_meta($job_id, '_featured');
    }

    // Clear Cache
    clean_post_cache($job_id);

    // Redirect to the job dashboard in the current language
    $dashboard_url = get_permalink(get_option('job_manager_job_dashboard_page_id'));
    if (function_exists('icl_get_current_language')) {
        $current_language = icl_get_current_language();
        $dashboard_url = apply_filters('wpml_permalink', $dashboard_url, $current_language);
    }

    wp_redirect(add_query_arg('msg', 'deactivated', $dashboard_url));
    exit;
}

// Success Message
add_action('job_manager_job_dashboard_before', 'custom_show_deactivation_notice');
function custom_show_deactivation_notice() {
    if (isset($_GET['msg']) && $_GET['msg'] === 'deactivated') {
        echo '<div class="job-manager-message success">' .
             esc_html__('Job successfully deactivated.', 'wp-job-manager') .
             '</div>';
    }
}

**Comment by workpei25** [comment]
Is this in the core plugin yet?

**Comment by Braehler** [comment]
@Scotm19 no this is not in the core plugin yet. Just put this into your functions.php

**Comment by workpei25** [comment]
I've tried that snippet but it's not working for me.

**Comment by Braehler** [comment]
This works currently on our live site without any problems with the lates WP Version and the latest WPJM 2.4.0 and WC Paid Listings 3.0.3. Be sure to have the wc paid listings add on active as well. And do not copy the starting <php

**Comment by workpei25** [comment]
Right, I'm not using Paid Listings. The action appears in the menu but has no effect when clicked.

**Comment by workpei25** [comment]
Actually, it does work, just not in Safari which supports an earlier support ticket I submitted about dashboard actions still not working in the dashboard. Thanks for the solution!
